Add "How Clerk works" and "Client handshake" to initial concepts section (#577)
clerk/clerk-docs: docs/concepts/client-handshake.mdx

---
title: Client handshake
description: The client handshake is a mechanism that is used to resolve a request’s authentication state from “unknown” to definitively signed in or signed out.
type: conceptual
---

> # The client handshake

Reviewer: Let's add a callout that this flow applies only to website browser traffic and not native apps.

> Clerk uses a client handshake mechanism to resolve a request’s authentication state from unknown (`handshake`) to `signed-in` or `signed-out`. Clerk’s session management architecture relies on a short-lived session JWT to validate requests, along with a long-lived session that keeps the session JWT fresh by interacting with the Frontend API while the web application is active. The long-lived session token is stored in an HttpOnly cookie associated with the Frontend API domain. If a short-lived session JWT is expired on a request to an application’s backend, the SDK doesn’t know if the session has ended, or if a new short-lived JWT needs to be issued. When an SDK gets into this state, it triggers the handshake.

Reviewer 2: ```suggestion
Clerk uses a mechanism called the handshake to resolve a request’s authentication state from unknown (`handshake`) to `signed-in` or `signed-out`. Clerk’s session management architecture relies on a short-lived session JWT to validate requests, along with a long-lived session that keeps the session JWT fresh by interacting with the Frontend API while the web application is active. The long-lived session token is stored in an HttpOnly cookie associated with the Frontend API domain. If a short-lived session JWT is expired on a request to an application’s backend, the SDK doesn’t know if the session has ended, or if a new short-lived JWT needs to be issued. When an SDK gets into this state, it triggers the handshake.
```

With the handshake, we can resolve authentication state on the backend and ensure the request is properly handled as signed in or out, instead of being in a potentially “unknown” state. The handshake flow relies on redirects to exchange session information between FAPI and the application, ensuring the resolution of unknown authentication states minimizes performance impact and behaves consistently across different framework and language implementations.

> The resolution is fast and shouldn't be noticeable to the end user.

Reviewer: Nit picky cooment: @BRKalow should we also mention that a part of it is handled by the Edge for marketing purposes?

Reviewer: @alexisintech 
```suggestion
The resolution is blazing fast and usually happens at the Edge.
```

## Handshake flow

The handshake mechanism operates by way of a redirect from the host application to a FAPI endpoint:

1. A request is made to an application using Clerk.
2. Clerk SDK determines the authentication state of the request (`signed-in`, `signed-out`, or `handshake`).

> 3. If the authentication state is `handshake`, Clerk responds with a 307 redirect to the handshake endpoint: `fapi/v1/client/handshake`.

Reviewer: ```suggestion
3. If the authentication state is `handshake`, Clerk SDK responds with a 307 redirect to the Frontend API handshake endpoint: `https://clerk.your-domain.com/v1/client/handshake`.
```

`https://clerk.your-domain.com` can be any FAPI test domain we use across the docs.

4. The handshake endpoint gets information about the current session and returns a handshake payload. The encoded handshake payload contains a list of `Set-Cookie` header directives to be passed along with the final response.
    - If the session is active, a fresh session JWT cookie is returned.
    - If the session is inactive, the session JWT cookie is wiped and the request will be treated as signed out.
5. The handshake endpoint redirects back to the host application along with the handshake payload, encoded either in the URL (development instances) or as a cookie (production instances).

> 6. The handshake payload is parsed and `Set-Cookie` headers are set on the response.

Reviewer: ```suggestion
6. The handshake payload is parsed by Clerk SDK and `Set-Cookie` headers are set on the response.
```

7. If an updated session JWT is returned, the JWT is verified and then, if verification is successful, the request is treated as signed in. If an updated session JWT is not returned, the request is treated as signed out.

## Scenarios that trigger a handshake

> 

Reviewer: A link to the basic concepts page when available in the section will be helpful as it explains the Clerk internals and the cookies.

<Callout type="warning">
A handshake is only triggered if a request is determined to be a document request (detected by the presence of the `Sec-Fetch-Dest: document` header, or `Accept: text/html`). Otherwise, the request is treated as `signed-out`.
</Callout>

1. Development instance and a dev browser is detected in URL 
[(URL-based session sync)](https://clerk.com/docs/upgrade-guides/url-based-session-syncing#url-based-session-syncing)
2. Production instance and request is being made to a satellite application

> 3. Has a session token (`__session`), but no active client is detected (`__client_uat` = 0 or missing)

Reviewer: ```suggestion
3. The current browser has a session token (`__session`), but no active client is detected (`__client_uat` = 0 or missing)
```

> 4. Has active client (`__client_uat` > 0) but no session token

Reviewer: ```suggestion
4. The current browser has active client (`__client_uat` > 0) but no session token
```

> 5. Client UAT is more recent than current session token issued at (`__client_uat` > `sessionToken.iat`)

Reviewer: ```suggestion
5. The `client_uat` cookie indicating the latest timestamp a client was updated, is more recent than current session token issued at timestamp (`__client_uat` > `sessionToken.iat`)
```

> 6. Session token exists but is expired

Reviewer: ```suggestion
6. The `__session` token cookie exists but is expired
```

> 7. Session token exists but is not active yet

Reviewer: ```suggestion
7. The `__session` token cookie exists but is not active yet
```

## Handshake architecture

The handshake flow is composed of two main pieces:

1. The FAPI [`/client/handshake` endpoint](https://clerk.com/docs/reference/frontend-api/tag/Client#operation/handshakeClient)

> 2. The `@clerk/backend` `authenticateRequest()` method

Reviewer: ```suggestion
2. The `@clerk/backend` `authenticateRequest()` method. The isomorphic `@clerk/backend` SDK is used by all framework specific Clerk SDKs such as `@clerk/nextjs`.
```

### Handshake payload

The handshake payload is a signed JWT that contains an array of `set-cookie` header directives. This allows us to transfer cookies from the FAPI domain to the application’s domain across environments without additional cookie setting logic embedded into our SDKs. In development, the payload is transported in the URL in a `__clerk_handshake` query parameter. In production, the payload is transported in a `__clerk_handshake` cookie. On returning from the handshake endpoint, the handshake payload is parsed and the cookie directives are set on the final response.

> 

Reviewer: @BRKalow Please add an example of a Handshake payload.

Reviewer: @alexisintech Please add the following example of the handshake payload:
```{
  "handshake": [
    "__client_uat=; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT",
    "__clerk_handshake=; Path=/; Domain=example.com; Expires=Thu, 01 Jan 1970 00:00:00 GMT",
    "__client_uat=1706822359; Path=/; Domain=example.com; Max-Age=315360000; Secure; SameSite=Lax",
    "__session=<...CLERK_SESSION_JWT...>; Path=/; Expires=Wed, 05 Feb 2025 09:08:53 GMT; Secure"
  ]
}```

---
title: Client handshake
description: The Client Handshake is a mechanism that is used to resolve a request’s authentication state from “unknown” to definitively signed in or signed out.
type: conceptual
---

# The client handshake

> Clerk uses a client handshake mechanism to resolve a request’s authentication state from unknown (`handshake`) to `signed-in` or `signed-out`. Clerk’s session management architecture relies on a short-lived session JWT to validate requests, along with a long-lived session that is used to keep the the session JWT fresh by interacting with the Frontend API. The long-lived session token is stored in an HttpOnly cookie associated with the Frontend API domain. If a short-lived session JWT is expired on a request to an application’s backend, the SDK doesn’t know if the session has ended, or if a new short-lived JWT needs to be issued. When an SDK gets into this state, it triggers the handshake.

Reviewer: ```suggestion
Clerk uses a client handshake mechanism to resolve a request’s authentication state from unknown (`handshake`) to `signed-in` or `signed-out`. Clerk’s session management architecture relies on a short-lived session JWT to validate requests, along with a long-lived session that keeps the session JWT fresh by interacting with the Frontend API while the web application is active. The long-lived session token is stored in an HttpOnly cookie associated with the Frontend API domain. If a short-lived session JWT is expired on a request to an application’s backend, the SDK doesn’t know if the session has ended, or if a new short-lived JWT needs to be issued. When an SDK gets into this state, it triggers the handshake.
```

> With the handshake, we can resolve authentication state on the backend and ensure the request is properly handled as signed in or out, instead of being in a potentially “unknown” state. The handshake flow relies on redirects to exchange session information between FAPI and the application, ensuring the resolution of unknown authentication states minimizes performance impact and behaves consistently across different framework and language implementations.

Reviewer: ```suggestion
With the handshake, we can resolve authentication state on the backend and ensure the request is properly handled as signed in or out, instead of being in a potentially “unknown” state. The handshake flow relies on redirects to exchange session information between FAPI and the application, ensuring the resolution of unknown authentication states minimizes performance impact and behaves consistently across different framework and language implementations.

The resolution is fast and shouldn't be noticeable to the end user.
```

8. If an updated session JWT is not returned, the request is treated as signed out

## Scenarios that trigger a handshake

<Callout type="warning">
A handshake is only triggered If a request is determined to be a document request (detected by the presence of the `Sec-Fetch-Dest: document` header, or `Accept: text/html`). Otherwise, the request is treated as `signed-out`.
</Callout>

1. Development instance and a dev browser is detected in URL **[URL-based session sync]**
2. Production instance and request is being made to a satellite application **[satellite needs sync]**
3. Has a session token (`__session`), but no active client is detected (`__client_uat` = 0 or missing) **[session token without client UAT]**
4. Has active client (`__client_uat` > 0) but no session token **[client UAT without session token]**
5. Client UAT is more recent than current session token issued at (`__client_uat` > `sessionToken.iat`) **[session token outdated]**
6. Session token exists but is expired **[session token expired]**
7. Session token exists but is not active yet **[session token not active yet]**

## Handshake architecture

The handshake flow is composed of two main pieces:

> 1. The FAPI [`/client/handshake` endpoint](https://github.com/clerk/clerk_go/blob/main/api/fapi/v1/router/router.go#L223)

Reviewer: Please replace this link that points to our internal Go repo with a link to the [FAPI API Reference](https://clerk.com/docs/reference/frontend-api/). 

The FAPI API reference doesn't contain the handshake endpoint yet but @Nikpolik can help us with that.

Reviewer 3: Sorry for this! PR to fix this https://github.com/clerk/clerk_go/pull/5243

> 2. The [`@clerk/backend` `authenticateRequest()` method](https://github.com/clerk/javascript/blob/main/packages/backend/src/tokens/request.ts#L74)

Reviewer: This method should be added to the SDK API reference docs (not automated yet). I'd avoid linking to a specific line in code as code changes, and the link will be rendered obsolete very soon.

Author: I'll remove both links for now then.

> ### `/client/handshake` endpoint

Reviewer: Please remove this section in favor of FAPI API Rerefence

Author: Remove the section? Or link to the FAPI reference? Or link to the handshake overview?

The endpoint should always result in a redirect back to the provided `redirect_url`. Along with the response, a **[handshake payload](#handshake-payload)** is returned. The handshake endpoint also handles multi-domain syncing between a satellite and a primary domain. If possible, the handshake will return a refreshed session token, as well as updated dev browser (`__clerk_db_jwt`) and Client UAT (`__client_uat`) cookies.

> ### `@clerk/backend` method

Reviewer: Ditto.

The `authenticateRequest()` method from the backend package primarily handles determining the authentication state of a given `Request` object. If the request state is determined to be `handshake`, the method returns returns headers to redirect to the handshake endpoint. These headers must then be set on the response returned by the SDK.

On redirect back to an application from the handshake endpoint, `authenticateRequest()` will run again and parse the handshake payload. Based on the contents of the payload, the request is determined to definitively be `signed-in` or `signed-out` and handled accordingly. The cookie directives from the handshake payload are set as `Set-Cookie` headers on the final response.

### Handshake payload

> The handshake payload is a signed JWT that contains an array of `set-cookie` header directives. This allows us to transfer cookies from the FAPI domain to the application’s domain across environments without additional cookie setting logic embedded into our SDKs. In development, the payload is transported in the URL in a `__clerk_handshake` query parameter. In production, the payload is transported in a `__clerk_handshake` cookie. On returning from the handshake endpoint, the handshake payload is parsed and the cookie directives are set on the final response.

Reviewer: ```suggestion
The handshake payload is a signed JWT that contains an array of `set-cookie` header directives. This allows us to transfer cookies from the FAPI domain to the application’s domain across environments without additional cookie setting logic embedded into our SDKs. In development, the payload is transported in the URL in a `__clerk_handshake` query parameter. In production, the payload is transported in a `__clerk_handshake` cookie. On returning from the handshake endpoint, the handshake payload is parsed and the cookie directives are set on the final response.

To verify the handshake payload JWT, use the instance [public RSA key](link to FAPI /.well-known/jwks.json)
```

Reviewer: The jwks link will be added by Nikos ASAP https://clerkinc.slack.com/archives/C066AN448GG/p1703255256891329
